Add algorithm profiling instrumentation and doc (#24103)
mantidproject/mantid: dev-docs/source/AlgorithmProfiler.rst

==============================
Work flows algorithm profiling
==============================

.. contents:: Contents
    :local:

Summary
^^^^^^^

> Due to the need of investigation of algorithms performance issues, the proper method

Reviewer: I suggest
"This describes how one can profile reduction workflows (CPU and RAM usage) that use one or more algorithms. The profiling tool requires two steps: the first is to build Mantid with a special CMake flag, and the second is to clone the visualization tool which is currently being kept as a separate project. Available for Linux only"

is introduced. It consists two to parts: special mantid build and analytical tool.

> Available for Linux only.

Reviewer: So is it only for Linux or can you use it on OSX also?

Mantid build
^^^^^^^^^^^^

> To build mantid for profiling run ``cmake`` with the additional option ``-DPROFILE_ALGORITHM_LINUX=ON``.

Reviewer: `-DALGORITHM_PROFILER=ON`?

Reviewer: "To build `Mantid` for..." ?

Author: Maybe the other flag would be needed for another os because of different profile functionality, so I'd leave LINUX in the flag. If somebody would implement something for other systems then would rename flag.

> Built in such a way mantid creates a dump file ``algotimeregister.out`` in the running directory.

Reviewer: "Built in such a way, `Mantid` will create..."

This file contains the time stamps for start and finish of executed algorithms with ~nanocecond
precision in a very simple text format.

Analysing tool
^^^^^^^^^^^^^^

The project is available here: https://github.com/nvaytet/mantid-profiler. It provides the nice graphical

> tool to interpret the information from dumped file.

Reviewer: "the information contained in the dumpe file."

Windows development
^^^^^^^^^^^^^^^^^^^

Precise timers are different for Linux and Windows (chrono is not good enough), so we need to treat them
separately. The suggestion is either to modify files ``Framework/API/inc/MantidAPI/AlgoTimeRegister.h`` and

> ``Framework/API/src/AlgoTimeRegister.cpp`` in manner ``#ifdef __WIN32`` or create the specific files with

Reviewer: ```suggestion
``Framework/API/src/AlgoTimeRegister.cpp`` by including ``#ifdef __WIN32``, or create the specific files with
```

the ``AlgoTimeRegister`` class for Windows and then in ``Framework/API/CMakeLists.txt`` edit the chunk that

> describes option PROFILE_ALGORITHM_LINUX combined with OS defied flags e.g., also in this case

Reviewer: ```suggestion
describes option PROFILE_ALGORITHM_LINUX combined with OS defined flags. In this case
```

> ``Framework/API/src/AlgorithmExecuteProfile.cpp`` needs to be modified or substituted.

Reviewer: ```suggestion
``Framework/API/src/AlgorithmExecuteProfile.cpp`` also needs to be modified or substituted.
```

> ``${CMAKE_SYSTEM_NAME} MATCHES "Windows"``. QueryPerformanceCounter supposed to be used on windows,

Reviewer: I'm not sure i understand this sentence.

Mantid build
^^^^^^^^^^^^

To build mantid version with profiling functionality enabled run ``cmake`` with the additional option
``-DPROFILE_ALGORITHM_LINUX=ON``. Built in such a way mantid creates a dump file ``algotimeregister.out``
in the running directory. This file contains the time stamps for start and finish of executed algorithms with

> ~nanocecond precision in a very simple text format.

Reviewer: nano**s**econd

Analysing tool
^^^^^^^^^^^^^^

The project is available here: https://github.com/nvaytet/mantid-profiler. It provides the nice graphical
tool to interpret the information contained in the dumped file.

Windows development
^^^^^^^^^^^^^^^^^^^

Precise timers are different for Linux and Windows (chrono is not good enough), so we need to treat them
separately. The suggestion is either to modify files ``Framework/API/inc/MantidAPI/AlgoTimeRegister.h`` and
``Framework/API/src/AlgoTimeRegister.cpp`` by including ``#ifdef __WIN32``, or create the specific files with
the ``AlgoTimeRegister`` class defined for Windows.  In the second case an inclusion of OS specific files should be
done correctly in ``Framework/API/CMakeLists.txt`  using ``${CMAKE_SYSTEM_NAME} MATCHES "Windows"``,
``Framework/API/src/AlgorithmExecuteProfile.cpp`` needs to be modified or substituted according to the type of OS.

> QueryPerformanceCounter supposed to be used on windows, instead of clock_gettime.

Reviewer: QueryPerformanceCounter **should** to be used on windows, instead of clock_gettime.

Reviewer: also add code formatting for "QueryPerformanceCounter" and "clock_gettime"?

To build mantid version with profiling functionality enabled run ``cmake`` with the additional option
``-DPROFILE_ALGORITHM_LINUX=ON``. Built in such a way mantid creates a dump file ``algotimeregister.out``
in the running directory. This file contains the time stamps for start and finish of executed algorithms with

> ~nanoceconds precision in a very simple text format.

Reviewer: nano**s**econd
